### ISSUE TITLE
[P0] C++/CUDA kernel integration gap — 11.5K lines of CUDA not wired into training path

### ISSUE THREAD

**Issue by dylanyunlon** [dylanyunlon/Neuron_SP#155]
## Problem

`deepspeed/runtime/desloc_engine.py` (3394 lines) does not call any of the 10 CUDA kernels in `csrc/hetero_reduce/` during the gradient synchronization path. The training loop uses `finalize_model_grads()` (Megatron DDP bucket all-reduce) for gradient sync and `torch.optim.AdamW(fused=True)` / `DeepSpeedCPUAdam` for the optimizer step — neither route invokes `fused_gradient_allreduce`, `hetero_reduce_scatter`, or `fused_adam_heterogeneous` from `csrc/hetero_reduce/`.

Concrete evidence:
- `deepspeed/runtime/desloc_engine.py:2515` — NaN check uses `dist.all_reduce(_nan_tensor, ...)` ✓ (control-flow only, acceptable)
- `deepspeed/runtime/desloc_engine.py:2677` — skip-tensor uses `dist.all_reduce(_skip_tensor, ...)` ✓ (control-flow only, acceptable)
- `deepspeed/runtime/desloc_engine.py:1286–1315` — gradient reduce-scatter delegated to `finalize_model_grads()` / DDP, **not** `hetero_reduce_scatter`
- `deepspeed/runtime/desloc_optimizer.py:58–95` — optimizer selects `torch.optim.AdamW(fused=True)` (VRAM ≥ 90 GB) or `DeepSpeedCPUAdam` (VRAM < 50 GB), **not** `fused_adam_heterogeneous`

## Root Cause

Three integration gaps exist:

**Gap 1 — gradient allreduce not using hetero kernels** (`deepspeed/runtime/desloc_engine.py:1286–1315`)  
`_setup_gradient_sync()` delegates to Megatron's `finalize_model_grads()` (DDP bucket all-reduce). The `HeteroReduceOp.fused_gradient_allreduce()` wrapper exists in `deepspeed/ops/hetero_reduce/hetero_reduce_op.py:467` but is never called.

**Gap 2 — optimizer not using per-tier Adam kernel** (`deepspeed/runtime/desloc_optimizer.py:51–116`)  
`DesLocOptimizerBuilder.build()` dispatches on VRAM thresholds (< 50 GB → CPUAdam, ≥ 90 GB → AdamW fused). `fused_adam_heterogeneous` from `csrc/hetero_reduce/fused_adam_heterogeneous.cu` (implements per-tier LR scaling via `HeteroReduceOp.fused_adam_heterogeneous()` at `deepspeed/ops/hetero_reduce/hetero_reduce_op.py:266`) is never selected.

**Gap 3 — op_builder not integrated into setup.py** (`op_builder/hetero_reduce.py`)  
`HeteroReduceBuilder` exists and `csrc/hetero_reduce/CMakeLists.txt` targets SM86/SM90/SM120 correctly, but `setup.py` / `op_builder/__init__.py` do not register `HeteroReduceBuilder` for pre-compilation during `pip install`.

### Gap 1 — wire hetero gradient allreduce (`deepspeed/runtime/desloc_engine.py`)

Replace the `finalize_model_grads()` call at line ~2575 with a hetero-aware path:

```python
# deepspeed/runtime/desloc_engine.py  (around line 2570)
try:
    from deepspeed.ops.hetero_reduce.hetero_reduce_op import HeteroReduceOp as _HROp
    _hetero_op = _HROp()
except Exception:
    _hetero_op = None

# In _gradient_sync() at line ~2575, replace finalize_model_grads block:
if _hetero_op is not None and not _fmg_skip_sync:
    _sm = 86 if torch.cuda.get_device_capability()[0] < 9 else (
          90 if torch.cuda.get_device_capability() < (12, 0) else 120)
    _hetero_op.fused_gradient_allreduce(
        grad=_flat_grad,
        int8_staging=_int8_staging,
        scale_staging=_scale_staging,
        ping_int8=_ping_int8, pong_int8=_pong_int8,
        ping_scale=_ping_scale, pong_scale=_pong_scale,
        rank=dist.get_rank(),
        world_size=dist.get_world_size(),
        sm_version=_sm,
    )
else:
    finalize_model_grads(
        [self._core_ddp if self._core_ddp is not None else self.model],
        num_tokens=None,
        skip_grad_sync=_fmg_skip_sync,
        force_all_reduce=False,
        pg_collection=_fmg_pg,
    )
```

### Gap 2 — wire fused_adam_heterogeneous (`deepspeed/runtime/desloc_optimizer.py`)

Add a new tier-aware path before the existing VRAM dispatch (line ~51):

```python
# deepspeed/runtime/desloc_optimizer.py  (insert before line 51)
try:
    from deepspeed.ops.hetero_reduce.hetero_reduce_op import HeteroReduceOp as _HROp
    _sm = 86 if torch.cuda.get_device_capability()[0] < 9 else (
          90 if torch.cuda.get_device_capability() < (12, 0) else 120)
    _hetero_adam = _HROp()
    # Wrap as a callable optimizer-compatible step
    class _HeteroAdamWrapper:
        def __init__(self, params, lr, betas, eps, weight_decay):
            self._params = list(params)
            self._lr = lr; self._betas = betas
            self._eps = eps; self._wd = weight_decay
            self._m = [torch.zeros_like(p) for p in self._params]
            self._v = [torch.zeros_like(p) for p in self._params]
            self._step = 0
        def step(self):
            self._step += 1
            for p, m, v in zip(self._params, self._m, self._v):
                if p.grad is None: continue
                _hetero_adam.fused_adam_heterogeneous(
                    param=p, grad=p.grad, exp_avg=m, exp_avg_sq=v,
                    step=self._step, lr=self._lr,
                    beta1=self._betas[0], beta2=self._betas[1],
                    eps=self._eps, weight_decay=self._wd, sm_version=_sm,
                )
    self.optimizer = _HeteroAdamWrapper(params, lr=cfg.lr,
        betas=(cfg.adam_beta1, cfg.adam_beta2),
        eps=cfg.adam_eps, weight_decay=cfg.weight_decay)
    self.optimizer_type = "hetero_adam_fused"
except Exception:
    pass  # fall through to existing VRAM-based dispatch
```

### Gap 3 — register builder (`op_builder/__init__.py`)

```python
# op_builder/__init__.py — add to ALL_OPS list
from .hetero_reduce import HeteroReduceBuilder
ALL_OPS["hetero_reduce"] = HeteroReduceBuilder
```

## Files

| File | Lines | Change |
|------|-------|--------|
| `deepspeed/runtime/desloc_engine.py` | 2565–2595 | Replace `finalize_model_grads` block with hetero allreduce dispatch |
| `deepspeed/runtime/desloc_optimizer.py` | 51–116 | Add `_HeteroAdamWrapper` tier-aware path before VRAM dispatch |
| `op_builder/__init__.py` | ALL_OPS dict | Register `HeteroReduceBuilder` |
| `deepspeed/ops/hetero_reduce/hetero_reduce_op.py` | existing | No changes needed — wrappers are complete (531 lines) |
| `csrc/hetero_reduce/CMakeLists.txt` | existing | No changes needed — SM86/SM90/SM120 targets present |

## Acceptance Criteria

```python
# 1. Import check
from deepspeed.ops.hetero_reduce import HeteroReduceOp
op = HeteroReduceOp()
assert hasattr(op, "fused_gradient_allreduce")
assert hasattr(op, "fused_adam_heterogeneous")

# 2. Optimizer type after Gap 2 fix
from deepspeed.runtime.desloc_optimizer import DesLocOptimizerBuilder
import types, torch
cfg = types.SimpleNamespace(lr=1e-4, adam_beta1=0.9, adam_beta2=0.999,
                             adam_eps=1e-8, weight_decay=0.1)
builder = DesLocOptimizerBuilder([torch.nn.Linear(64,64).cuda()], cfg)
assert builder.optimizer_type == "hetero_adam_fused", builder.optimizer_type

# 3. Gradient sync uses hetero path (integration test)
# Run: python -m pytest tests/unit/runtime/test_desloc_engine.py::test_hetero_grad_sync -v
# Expected: "hetero_reduce" appears in captured log, "finalize_model_grads" does NOT

# 4. Compile check
# cd csrc/hetero_reduce && mkdir -p build && cd build
# cmake .. -DCMAKE_CUDA_ARCHITECTURES="86;90;120" && make -j$(nproc)
# Expected: zero errors, produces libhetero_reduce_kernels.a
```

`benchmark_mfu.py` with `--use-hetero-kernels` flag should show ≥5% throughput improvement on the 2×A6000+H100 cluster vs. baseline `finalize_model_grads` path (measured in tokens/sec/GPU).

## Corrections to Original PRD

- ~~`desloc_engine.py` (root)~~ → correct path: `deepspeed/runtime/desloc_engine.py`
- ~~`deepspeed/ops/hetero_reduce/hetero_reduce_op.py` is incomplete~~ → file is complete (531 lines, all 10 kernels wrapped); no work needed there
- ~~`deepspeed/core/transformer/mlp.py` needs fused_swiglu_ln wiring~~ → already wired at lines 73–95
- ~~`deepspeed/core/transformer/attention.py` needs fused_rope_hetero wiring~~ → already wired at lines 222–277
- ~~Add `CMakeLists.txt`~~ → `csrc/hetero_reduce/CMakeLists.txt` already exists with SM86/90/120
